src/detect-flow-pkts.c

```diff
@@ -23,238 +23,412 @@
+        DetectFlowBytesFree(de_ctx, df);
         return -1;
     }
     s->flags |= SIG_FLAG_REQUIRE_PACKET;
 
     return 0;
 }
 
-void DetectFlowBytesToClientRegister(void)
+static int DetectFlowBytesToClientSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
 {
-    sigmatch_table[DETECT_FLOW_BYTES_TO_CLIENT].name = "flow.bytes_toclient";
-    sigmatch_table[DETECT_FLOW_BYTES_TO_CLIENT].desc = "match flow number of bytes to client";
-    sigmatch_table[DETECT_FLOW_BYTES_TO_CLIENT].url =
-            "/rules/flow-keywords.html#flow-bytes_toclient";
-    sigmatch_table[DETECT_FLOW_BYTES_TO_CLIENT].Match = DetectFlowBytesToClientMatch;
-    sigmatch_table[DETECT_FLOW_BYTES_TO_CLIENT].Setup = DetectFlowBytesToClientSetup;
-    sigmatch_table[DETECT_FLOW_BYTES_TO_CLIENT].Free = DetectFlowBytesToClientFree;
-}
+    DetectU64Data *du64 = DetectU64Parse(rawstr);
+    if (du64 == NULL)
+        return -1;
 
-static int DetectFlowBytesToServerMatch(
-        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
-{
-    if (p->flow == NULL) {
-        return 0;
-    }
-    uint64_t nb = p->flow->todstbytecnt;
+    DetectFlow *df = SCCalloc(1, sizeof(DetectFlow));
+    if (df == NULL)
+        return -1;
 
-    const DetectU64Data *du64 = (const DetectU64Data *)ctx;
-    return DetectU64Match(nb, du64);
-}
+    df->byte_data = du64;
+    df->dir = DETECT_FLOW_TOCLIENT;
 
-static void DetectFlowBytesToServerFree(DetectEngineCtx *de_ctx, void *ptr)
-{
-    rs_detect_u64_free(ptr);
+    if (SigMatchAppendSMToList(
+                de_ctx, s, DETECT_FLOW_BYTES, (SigMatchCtx *)df, DETECT_SM_LIST_MATCH) == NULL) {
+        DetectFlowBytesFree(de_ctx, df);
+        return -1;
+    }
+    s->flags |= SIG_FLAG_REQUIRE_PACKET;
+
+    return 0;
 }
 
-static int DetectFlowBytesToServerSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+static int DetectFlowBytesSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
 {
-    DetectU64Data *du64 = DetectU64Parse(rawstr);
-    if (du64 == NULL)
-        return -1;
+    DetectFlow *df = NULL;
+    char copy[strlen(rawstr) + 1];
+    strlcpy(copy, rawstr, sizeof(copy));
+    char *context = NULL;
+    char *token = strtok_r(copy, ",", &context);
+    uint8_t num_tokens = 1;
+    uint8_t dir = 0;
+    char *byte_data = NULL;
+
+    while (token != NULL) {
+        if (num_tokens > 2)
+            goto error;
+
+        while (*token != '\0' && isblank(*token)) {
+            token++;
+        }
+        if (strlen(token) == 0) {
+            goto next;
+        }
+
+        num_tokens++;
+
+        if (strcmp(token, "toserver") == 0) {
+            dir = DETECT_FLOW_TOSERVER;
+        } else if (strcmp(token, "toclient") == 0) {
+            dir = DETECT_FLOW_TOCLIENT;
+        } else if (strcmp(token, "either") == 0) {
+            dir = DETECT_FLOW_TOEITHER;
+        }
```

Comment:
should we have a else here? Seems a bit open ended like this

Reply:
Keeping toserver as the default

Reply:
hmm if i do that, any random token will result in toserver. It won't be strict..